File: go_scripts/pkg/qa/audit_upstream_links.go
package qa

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/nyamiiko/script_hub/go_scripts/pkg/hub"
)

func extractUrls(filePath string) []string {
	content := hub.ReadFileString(filePath)
	re := regexp.MustCompile(`https?://[^\s'"<>]+`)
	return re.FindAllString(content, -1)
}

func checkUrl(urlStr string) (string, int, string) {
	client := &http.Client{
		Timeout: 10 * time.Second,
		// Using default transport is fine here, skipping SSL validation isn't strictly necessary
		// unless there's a specific issue, but we can do a simple GET.
	}
	req, _ := http.NewRequest("GET", urlStr, nil)
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := client.Do(req)
	if err != nil {
		return urlStr, 0, err.Error()
	}
	defer resp.Body.Close()
	return urlStr, resp.StatusCode, ""
}

func RunAuditUpstreamLinks() int {
	pipelineDir := filepath.Join(hub.ROOT, "go_scripts/pkg/pipeline")
	fmt.Printf("Scanning for URLs in %s...\n", pipelineDir)

	allUrls := make(map[string]bool)
	err := filepath.Walk(pipelineDir, func(path string, info os.FileInfo, err error) error {
		if err == nil && !info.IsDir() && strings.HasSuffix(path, ".go") {
			urls := extractUrls(path)
			for _, u := range urls {
				if strings.Contains(u, "github.com") || strings.Contains(u, "raw.githubusercontent") || strings.Contains(u, "moe") || strings.Contains(u, "jsdelivr") || strings.Contains(u, "kelee") || strings.Contains(u, "yfamilys") {
					allUrls[u] = true
				}
			}
		}
		return nil
	})

	if err != nil {
		fmt.Printf("Error walking directory: %v\n", err)
		return 1
	}

	var urls []string
	for u := range allUrls {
		urls = append(urls, u)
	}
	sort.Strings(urls)
	fmt.Printf("Found %d unique upstream URLs to check.\n", len(urls))

	var failed []string
	for _, u := range urls {
		// Strictly serial execution to avoid rate limits / concurrency issues on network calls per user constraint
		urlStr, status, errStr := checkUrl(u)
		if status != 200 && status != 301 && status != 302 && status != 304 {
			fmt.Printf("[FAIL] %d | %s | %s\n", status, urlStr, errStr)
			failed = append(failed, urlStr)
		}
		time.Sleep(100 * time.Millisecond) // sleep to prevent rate limiting
	}

	if len(failed) > 0 {
		fmt.Printf("\n%d URLs failed.\n", len(failed))
		return 1
	}

	fmt.Println("\nAll URLs are accessible (200 OK)!")
	return 0
}

File: create/scripts/qa/audit_upstream_links.py
#!/usr/bin/env python3
import os
import re
import ssl
import urllib.request
import urllib.error
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor

SCRIPTS_DIR = Path(__file__).resolve().parents[1]
PIPELINE_DIR = SCRIPTS_DIR / "pipeline"

def extract_urls(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    # Simple regex for http/https URLs inside quotes
    urls = re.findall(r'https?://[^\s\'"<>]+', content)
    return urls

def check_url(url):
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
        # Using GET because some CDNs/GitHub block HEAD or return 403 for HEAD
        with urllib.request.urlopen(req, context=ctx, timeout=10) as response:
            return url, response.status, None
    except urllib.error.HTTPError as e:
        return url, e.code, str(e)
    except urllib.error.URLError as e:
        return url, 0, str(e.reason)
    except Exception as e:
        return url, 0, str(e)

def main():
    print(f"Scanning for URLs in {PIPELINE_DIR}...")
    all_urls = set()
    for root, _, files in os.walk(PIPELINE_DIR):
        for file in files:
            if file.endswith(".py"):
                path = Path(root) / file
                urls = extract_urls(path)
                for u in urls:
                    if "github.com" in u or "raw.githubusercontent" in u or "moe" in u or "jsdelivr" in u or "kelee" in u or "yfamilys" in u:
                        all_urls.add(u)
                        
    urls = sorted(list(all_urls))
    print(f"Found {len(urls)} unique upstream URLs to check.")
    
    failed = []
    with ThreadPoolExecutor(max_workers=10) as executor:
        results = executor.map(check_url, urls)
        for url, status, error in results:
            if status not in (200, 301, 302, 304):
                print(f"[FAIL] {status} | {url} | {error}")
                failed.append((url, status, error))
            else:
                # print(f"[OK] {status} | {url}")
                pass
                
    if failed:
        print(f"\n{len(failed)} URLs failed.")
        exit(1)
    else:
        print("\nAll URLs are accessible (200 OK)!")

if __name__ == "__main__":
    main()
